File: 0067-add-binary/0067-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i = len(a) - 1
        j = len(b) - 1
        cin = 0
        result = []

        while (i >= 0 or j >= 0 or cin):
            bitA = int(a[i]) if i >= 0 else 0
            bitB = int(b[j]) if j >= 0 else 0
            sum = bitA ^ bitB ^ cin
            cin = (bitA & bitB) | (cin & (bitA ^ bitB))
            result.append(str(sum))
            i -= 1
            j -= 1
        
        return ''.join(reversed(result))
        

File: 0067-add-binary/0067-add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int i = a.size() - 1;
        int j = b.size() - 1;
        int cin = 0;

        std::string result = "";

        while (i >= 0 || j >= 0 || cin) {
            int bitA = i >= 0 ? a[i] - '0' : 0;
            int bitB = j >= 0 ? b[j] - '0' : 0;
            int sum = bitA ^ bitB ^ cin;
            cin = (bitA & bitB) | (cin & (bitA ^ bitB));
            result += std::to_string(sum);

            --i;
            --j;
        }

        reverse(result.begin(), result.end());
        return result;
    }
};